### ISSUE TITLE
[CA] San Bernardino County

### ISSUE THREAD

**Issue by JeffreyBLewis** [kuriwaki/cvr_harvard-mit_scripts#240]
San Bernardino County has two basic issues.    First, the Harvard file is not correctly marking out overvotes and other records that are included in the raw data that were not counted in the final vote tally (records flagged as "ambiguous" or not "current").  Jim is rebuilding the long dta to address this and then @kuriwaki can rebuild the Harvard parquet.

The second issue is that small vote tally discrepancies of unknown origin will still exist after the first issue is addressed.   I will continue to try track the source of this down.   Many precincts match (at least for president) exactly, but in a minority, of precincts, there are discrepancies of a few votes.

**Comment by kuriwaki** [comment]
@mreece13 can you run your `preprocess_json` on your San Bernardino file only and tell me the number of Biden  and undervotes you get? (`tbl > count(candidate, PartyID, party_detailed, sort = TRUE)` after filtering on `contest == "US PRESIDENT`)

Your modifications [here](https://github.com/kuriwaki/cvr_harvard-mit_scripts/blob/800d8dfb382af77574604bfc747d8a7efe85f8ee/code/build-medsl/code/functions.R#L351) should correctly treat the undervotes just like a candidate. But there are no undervotes in your final parquet. I am working with the raw json files independently to figure out why.

**Comment by mreece13** [comment]
![image](https://github.com/user-attachments/assets/ab32e052-54c0-4628-a503-e891cd7b54ac)

@kuriwaki `pass0` does not do much formatting, so apologies for the somewhat messy output

**Comment by kuriwaki** [comment]
@mreece13 thanks, those candidate votes are ballpark correct (https://results.rov.sbcounty.gov/results/20201103/) but there should be 9445 undervotes in this contest, not 493. I get 9460 undervotes with my code below that uses some of your filters. Thoughts on why yours is low? Also, even those 493 undervotes seem to be getting removed at some subsequent point. 


Here is my code with my parser

```
library(tidyverse)
library(fs)
library(dominionCVR) # kuriwaki/dominionCVR

# raw files
# https://votedatabase.com/Cvr/California/San%20Bernardino/
# CVR_Export_20220914135729 (1) K Call - San Bernardino, CA.zip from 
cvr_files <- dir_ls("~/Downloads/sb/", regexp = "CvrExport")

# read_manifest("~/Downloads/sb/CandidateManifest.json")
cand_Ids <- tibble::tribble(
         ~candidate, ~candidateId,      ~candType,
   "JOSEPH R. BIDEN",      1L,          "Regular",
   "DONALD J. TRUMP",      2L,          "Regular",
    "GLORIA LA RIVA",      3L,          "Regular",
"ROCKY DE LA FUENTE",      5L,          "Regular",
     "HOWIE HAWKINS",      4L,          "Regular",
      "JO JORGENSEN",      6L,          "Regular",
          "Write-in",    543L,          "WriteIn",
     "BRIAN CARROLL",    625L, "QualifiedWriteIn",
      "MARK CHARLES",    626L, "QualifiedWriteIn",
    "JOSEPH KISHORE",    627L, "QualifiedWriteIn",
      "BROCK PIERCE",    628L, "QualifiedWriteIn",
     "JESSE VENTURA",    629L, "QualifiedWriteIn"
           )

# multicore read (20secs)
cvrexport_all <- extract_cvr(path = cvr_files) |>
  as_tibble()

# President only
cvr_pres <- cvrexport_all |>
  filter(contestId == 1) |>
  filter(isCurrent) |>
  select(!originalModified) |>
  left_join(cand_Ids, by = c("candidateId"), relationship = "many-to-one")

# vote counts for president
# https://github.com/kuriwaki/cvr_harvard-mit_scripts/blob/800d8dfb382af77574604bfc747d8a7efe85f8ee/code/build-medsl/code/functions.R#L351
cvr_pres |>
  filter((isVote | undervotes > 0 | overvotes > 0) | (!isAmbiguous & undervotes == 0 & overvotes == 0)) |>
  mutate(
    candidate = case_when(
      undervotes > 0 ~ "UNDERVOTE",
      overvotes > 0 ~ "OVERVOTE",
      .default = candidate
  )) |>
  count(candidate, sort = TRUE)
```
```
   candidate                                  n
   <chr>                                  <int>
 1 "JOSEPH R. BIDEN"                     455710
 2 "DONALD J. TRUMP"                     365962
 3 "UNDERVOTE"                             9460
 4 "JO JORGENSEN"                          8783
 5 "OVERVOTE"                              5287
 6 "HOWIE HAWKINS"                         3619
```

**Comment by mreece13** [comment]
Hmmm, I went and looked through the `dominionCVR` repo and there are not obvious differences to me (I am also not that fluent in C++). I think there are differences in how I distinguish between the `Modified` cards and the `Original` cards, versus how `dominionCVR` does. 

Since the estimates from the library are correct, I'll rework the `preprocess_json()` code to use that library instead. We probably need to make the repo public so that we can pass replication.

**Comment by kuriwaki** [comment]
I don't think the inside of the `dominionCVR::extract_cvr` is that special, in that were not modifying any values or removing rows in that function. I was not thinking the discrepancy comes from there, though you never know. 

Yes, in Arizona it seems like we filtered to `isCurrent`, which removes multiple attempts at the same contest (this is what seems to created the `Original` and `Modified`). I did that outside of dominionCVR::extract_cvr() though.
